src/components/MessageList/MessageList.tsx

```diff
@@ -825,36 +830,57 @@ const MessageListWithContext = <
   const onScrollEndDrag = () =>
     hasMoved && selectedPicker && setHasMoved(false);
 
+  const refCallback = (
+    ref: FlatListType<MessageType<At, Ch, Co, Ev, Me, Re, Us>>,
+  ) => {
+    flatListRef.current = ref;
+
+    if (setFlatListRef) {
+      setFlatListRef(ref);
+    }
+  };
+
+  const renderFooterComponent = () => (
+    <FooterComponent loadingMore={onEndReachedInProgress} />
+  );
+
+  const renderHeaderComponent = () => (
+    <HeaderComponent loadingMore={onStartReachedInProgress} />
+  );
+
+  const renderEmptyStateIndicator = () => (
```

Comment:
So rendering empty state is still in flatlist, I have just separated out inline fuction. Inline functions can cause more re-renders. React documentation only gives headsup about this for ref callback function (inline vs separate function), it doesn't say any such thing about indicators. But from the discussions I can see, it's still recommended to not have inline functions in FlatList.

For loading indicator, I had to take it out of FlatList because
- when you are in network recovery state, we show loading indicator.
- after recovery, when new results are ready, scroll doesn't start at bottom of the list if loading indicator is part of FlatList (within ListEmptyComponent).

And this seemed like a confusing UX to me.